src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs

```diff
@@ -75,6 +75,11 @@ public async Task SynchronizeWithServer()
             }
             try
             {
+                if (!sonarQubeService.IsConnected)
+                {
+                    return;
+                }
```

Comment:
The fetch attempt will only happen if the binding configuration is "Connected Mode". So if we got to this point, we should be in connected mode yet SQ client might not be initialized yet.

This code can be called as a response to 2 events:

1. [ActiveSolutionBoundTracker.SolutionBindingChanged](https://github.com/SonarSource/sonarlint-visualstudio/blob/master/src/Integration/MefServices/ActiveSolutionBoundTracker.cs#L143) event.
This event is raised in 2 use cases:
1.1 When the active solution is changed [IActiveSolutionTracker.ActiveSolutionChanged](https://github.com/SonarSource/sonarlint-visualstudio/blob/master/src/Integration/MefServices/ActiveSolutionBoundTracker.cs#L78). If the opened solution has connected mode configured, `SolutionBindingChanged` event will always be preceded with [awaiting a connection to SQ server](https://github.com/SonarSource/sonarlint-visualstudio/blob/4813fca6e9133af9b64c3c1171f878c5f6c55003/src/Integration/MefServices/ActiveSolutionBoundTracker.cs#L123).
So in this case, TaintIssuesSynchronizer will always succeed.

1.2 When team explorer SQ pane is first initialized/changed (https://github.com/SonarSource/sonarlint-visualstudio/blob/4813fca6e9133af9b64c3c1171f878c5f6c55003/src/Integration/MefServices/ActiveSolutionBoundTracker.cs#L75). 
It originates in StateManager.SetProjects, which is called by the team explorer pane's controller.
In this case, TaintIssuesSynchronizer will always fail, since the SQ server is not yet initialized at this point.

1.1 and 1.2 have an interaction between them: if the team explorer pane is already open when a bound solution is opened, 1.2 will always win and 1.1 will never be called due to [this IF](https://github.com/SonarSource/sonarlint-visualstudio/blob/4813fca6e9133af9b64c3c1171f878c5f6c55003/src/Integration/MefServices/ActiveSolutionBoundTracker.cs#L140).

2. [ActiveSolutionBoundTracker.SolutionBindingUpdated](https://github.com/SonarSource/sonarlint-visualstudio/blob/4813fca6e9133af9b64c3c1171f878c5f6c55003/src/Integration/MefServices/ActiveSolutionBoundTracker.cs#L147).
This is always called by the team explorer SQ pane's controller. If the pane is never opened, it will never be called.

So unless I missed something, it seems that the execution would depend on whether or not the user has opened the team explorer pane. 
If they did, we always fail on `SolutionBindingChanged` event and rely on `SolutionBindingUpdated` to be raised. 
If the pane has not been opened, then only `SolutionBindingChanged` will be called and it would always be preceded with awaiting the connection to the server.
So it seems that checking if the server is connected is enough to handle the two cases, but I must admit that I'm worried that I missed something.

I've also tried to check what happens if a user manually opens the tool window during the solution load (e.g. after opening the solution but before `SolutionBindingChanged` is raised). However that doesn't seem to be possible: `IActiveSolutionTracker.ActiveSolutionChanged` is raised on the UI thread, and it immediately [disconnects the SQ server](https://github.com/SonarSource/sonarlint-visualstudio/blob/4813fca6e9133af9b64c3c1171f878c5f6c55003/src/Integration/MefServices/ActiveSolutionBoundTracker.cs#L103). So I can't seem to open the taint tool window after the solution is opened and yet before this block of code is executed - either the tool window is opened before the solution is loaded (in which case we are in standalone mode), or it is opened after VS raises this event and SQ is already disconnected and `SolutionBindingChanged` will be fired later. 

I tried to reproduce various cases and this block of code seems to work but this entire thing feels wrong.
Why do we even need to wait for the SQ client to be "connected"? We have the credentials, and every call should be stateless and just try to fetch the data. If it fails then it fails. I don't see why we would need a global IsConnected flag that gives us so much trouble.

Reply:
As discussed offline, I suggest opening a ticket about the sync order and copy your notes above into it.